File: web/src/api/shares.ts
import { request } from './client'

export interface ShareInfo {
  id: string
  file_id: string | null
  folder_id: string | null
  token: string
  has_password: boolean
  permission: string
  max_downloads: number | null
  download_count: number
  expires_at: string | null
  created_at: string
}

export interface PublicShareInfo {
  token: string
  has_password: boolean
  file_name: string | null
  file_size: number | null
  expires_at: string | null
}

export interface CreateShareRequest {
  file_id?: string
  folder_id?: string
  password?: string
  max_downloads?: number
  expires_at?: string
}

export async function createShare(data: CreateShareRequest): Promise<ShareInfo> {
  return request<ShareInfo>('/shares', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(data),
  })
}

export async function listShares(): Promise<ShareInfo[]> {
  return request<ShareInfo[]>('/shares')
}

export async function deleteShare(id: string): Promise<void> {
  return request<void>(`/shares/${id}`, { method: 'DELETE' })
}

// Public share endpoints (no auth)
export async function getPublicShare(token: string): Promise<PublicShareInfo> {
  const res = await fetch(`/api/v1/shares/public/${token}`)
  if (!res.ok) {
    const body = await res.json()
    throw new Error(body.error?.message || '分享不存在')
  }
  const json = await res.json()
  return json.data
}

export async function verifySharePassword(token: string, password: string): Promise<PublicShareInfo> {
  const res = await fetch(`/api/v1/shares/public/${token}/verify`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ password }),
  })
  if (!res.ok) {
    const body = await res.json()
    throw new Error(body.error?.message || '密码错误')
  }
  const json = await res.json()
  return json.data
}

export function getShareDownloadUrl(token: string): string {
  return `/api/v1/shares/public/${token}/download`
}

File: crates/api/src/shares.rs
use axum::{
    Json, Router,
    extract::{Path, State},
    http::{StatusCode, header},
    response::IntoResponse,
    routing::{get, post},
};
use mdp_auth::middleware::AuthUser;
use mdp_common::{error::AppError, response::ApiResponse};
use mdp_core::share::{PublicShareInfo, ShareInfo, ShareService};
use serde::Deserialize;
use uuid::Uuid;

use crate::state::AppState;

/// Authenticated share management routes: /api/v1/shares
pub fn routes() -> Router<AppState> {
    Router::new()
        .route("/shares", post(create_share).get(list_shares))
        .route("/shares/{id}", axum::routing::delete(delete_share))
}

/// Public share access routes (no auth required)
pub fn public_routes() -> Router<AppState> {
    Router::new()
        .route("/shares/public/{token}", get(get_share))
        .route("/shares/public/{token}/verify", post(verify_share))
        .route("/shares/public/{token}/download", get(download_share))
}

#[derive(Deserialize)]
struct CreateShareRequest {
    file_id: Option<Uuid>,
    folder_id: Option<Uuid>,
    password: Option<String>,
    max_downloads: Option<i32>,
    /// ISO 8601 datetime
    expires_at: Option<chrono::DateTime<chrono::Utc>>,
}

async fn create_share(
    State(state): State<AppState>,
    auth: AuthUser,
    Json(req): Json<CreateShareRequest>,
) -> Result<ApiResponse<ShareInfo>, AppError> {
    let share = ShareService::create(
        &state.db,
        auth.user_id,
        req.file_id,
        req.folder_id,
        req.password.as_deref(),
        req.max_downloads,
        req.expires_at,
    )
    .await?;
    Ok(ApiResponse::ok(share))
}

async fn list_shares(
    State(state): State<AppState>,
    auth: AuthUser,
) -> Result<ApiResponse<Vec<ShareInfo>>, AppError> {
    let shares = ShareService::list_by_user(&state.db, auth.user_id).await?;
    Ok(ApiResponse::ok(shares))
}

async fn delete_share(
    State(state): State<AppState>,
    auth: AuthUser,
    Path(id): Path<Uuid>,
) -> Result<StatusCode, AppError> {
    ShareService::delete(&state.db, auth.user_id, id).await?;
    Ok(StatusCode::NO_CONTENT)
}

async fn get_share(
    State(state): State<AppState>,
    Path(token): Path<String>,
) -> Result<ApiResponse<PublicShareInfo>, AppError> {
    let (_, info) = ShareService::get_by_token(&state.db, &token).await?;
    Ok(ApiResponse::ok(info))
}

#[derive(Deserialize)]
struct VerifyRequest {
    password: String,
}

async fn verify_share(
    State(state): State<AppState>,
    Path(token): Path<String>,
    Json(req): Json<VerifyRequest>,
) -> Result<ApiResponse<PublicShareInfo>, AppError> {
    let (share, info) = ShareService::get_by_token(&state.db, &token).await?;
    if !ShareService::verify_password(&share, &req.password) {
        return Err(AppError::Forbidden);
    }
    Ok(ApiResponse::ok(info))
}

async fn download_share(
    State(state): State<AppState>,
    Path(token): Path<String>,
) -> Result<impl IntoResponse, AppError> {
    let (share, _) = ShareService::get_by_token(&state.db, &token).await?;

    // If share has a password, require it via query param or deny
    if share.password.is_some() {
        return Err(AppError::Forbidden);
    }

    let (file, data) = ShareService::download(&state.db, &state.storage, share).await?;

    let headers = [
        (header::CONTENT_TYPE, file.content_type),
        (
            header::CONTENT_DISPOSITION,
            format!("attachment; filename=\"{}\"", file.name),
        ),
    ];

    Ok((StatusCode::OK, headers, data))
}
